### DISCUSSION TITLE
Better type targeted new inference

### DISCUSSION THREAD

**Discussion by NN---** [dotnet/csharplang#5998] [Language Ideas]
The compiler cannot infer return generic type from type targeted new.

```cs
using System;

Either<L, W> FEither<L, R, W>(
    Either<L, R> either,
    Func<R, Either<L, W>> f) =>
    either.Left is null ? f(either.Right) : new(either.Left);

Either<Exception, double> Add(Either<Exception, double> a, double b) =>
    FEither(a, r => new Either<Exception, double>(r + b)); // OK

Either<Exception, double> Sub(Either<Exception, double> a, double b) =>
    FEither(a, r => new(r - b)); // Doesn't compile, must specify type explicitly

public record Either<L, R>
{
    public Either(L l) => Left = l;
    public Either(R r) => Right = r;

    public L Left;
    public R Right;
}
```

**Comment by Rekkonnect** [comment]
If I had to guess from this example, it is not target-typing the lambda's signature to the delegate type, thus not propagating to the target-typed `new`. In other words, it does not try to make `r => new(r - b)` a `Func<double, Either<Exception, double>>`, so it cannot target-type `new(r - b)` to an `Either<Exception, double>`.

Therefore, this will require improvements on the lambda's side of target-typing.
